File: python/mgBinL.py
""" mgBinL.py (since written in a functional style, annotated for mypy) """
from mgBin import *

def negWS(ws) -> bool:
  return [lso for lso in fset.toList(ws) if lso[1][0] != ()] != []

def ell(so) -> frozenset:
  """ map so to its derived workspace """
  if isinstance(so,tuple) and len(so) == 2 and \
     isinstance(so[0],tuple) and ( len(so[0]) == 0 or isinstance(so[0][0],str) ): # lex
    return fset.fromList([(so,so[1])])
  elif isinstance(so,frozenset): # set
    ([nws],[pws]) = partition(negWS, [ell(s) for s in fset.toList(so)])
    (nlsos,others) = partition( lambda lso: lso[1][0] != (), fset.toList(nws) )
    if len(nlsos) != 1:
      raise RuntimeError("ell: zero or >1 neg LSOs in workspaces")
    (nso,(nfs,pfs)) = nlsos[0]
    imatches = [lso for lso in others if lso[1][1][0] == nfs[0]]
    if len(imatches)==1:
      (pso,plabel) = imatches[0]
      if maxx(fset.toList(pws))[0] != pso:
        raise RuntimeError("move-over-merge error")
      else: # im
        return d([nws])
    else: # em
      return d([nws,pws])
  else:
    raise TypeError("ell type error")

File: haskell/MgBinL.hs
module MgBinTransduction where
import Data.Set (Set)
import qualified Data.Set as Set
import Data.List
import qualified Data.List as List

import MgBin

negWS :: WS -> Bool
negWS ws = fst (Set.partition ((/= []).fst.snd) ws) /= (Set.empty)

ell :: SO -> WS
ell (L (w,l)) = (Set.singleton (L (w,l), l))
ell (S s) = case List.partition negWS (map ell (Set.toList s)) of
  ([nws],[pws]) -> case List.partition ((/= []).fst.snd) (Set.toList nws) of
      ([(nso,(f:ns,ps))],others) -> case List.filter ((== f).head.snd.snd) others of
          [(pso,plabel)] -> if (fst.maxx) (Set.toList pws) /= pso -- IM
                            then (error "ell: move-over-merge violation")
                            else d [nws]
          [] -> d [nws,pws] -- EM

